Add torch.library.custom_op

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #122796
* #123129
* #123110
* #123109
* #123108
* #123107
* __->__ #122344

This is the entrypoint for defining an opaque/blackbox (e.g. PyTorch will
never peek into it) custom op. In this PR, you can specify backend impls
and the abstract impl for this op.

NB: most of this PR is docstrings, please don't be intimidated by the
line count.

There are a number of interesting features:
- we infer the schema from type hints. In a followup I add the ability
  to manually specify a schema.
- name inference. The user needs to manually specify an op name for now.
  In a followup we add the ability to automatically infer a name (this
  is a little tricky).
- custom_op registrations can override each other. This makes them
  more pleasant to work with in environments like colab.
- we require that the outputs of the custom_op do not alias any inputs
  or each other. We enforce this via a runtime check, but can relax this
  into an opcheck test if it really matters in the future.

Test Plan:
- new tests

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang

zou3519 committed:
Add torch.library.def_blackbox

This is the entrypoint for defining a blackbox (e.g. PyTorch will
never peek into it) custom op. In this PR, you can specify backend impls
and the abstract impl for this op.

NB: most of this PR is docstrings, please don't be intimidated by the
line count.

There are a number of interesting features:
- we infer the schema from type hints. In a followup I add the ability
  to manually specify a schema.
- we infer the namespace for the operator from the module. In a
  followup, we'll add the ability to manually specify a namespace.
- def_blackbox registrations can override each other. This makes them
  more pleasant to work with in environments like colab.
- we require that the outputs of the blackbox op do not alias any inputs
  or each other. We enforce this via a runtime check.

Test Plan:
- new tests

[ghstack-poisoned]

zou3519 committed:
Update on "Add torch.library.def_blackbox"


This is the entrypoint for defining a blackbox (e.g. PyTorch will
never peek into it) custom op. In this PR, you can specify backend impls
and the abstract impl for this op.

NB: most of this PR is docstrings, please don't be intimidated by the
line count.

There are a number of interesting features:
- we infer the schema from type hints. In a followup I add the ability
  to manually specify a schema.
- we infer the namespace for the operator from the module. In a
  followup, we'll add the ability to manually specify a namespace.
- def_blackbox registrations can override each other. This makes them
  more pleasant to work with in environments like colab.
- we require that the outputs of the blackbox op do not alias any inputs
  or each other. We enforce this via a runtime check.

Test Plan:
- new tests

[ghstack-poisoned]

zou3519 committed:
Update on "Add torch.library.def_blackbox"


This is the entrypoint for defining a blackbox (e.g. PyTorch will
never peek into it) custom op. In this PR, you can specify backend impls
and the abstract impl for this op.

NB: most of this PR is docstrings, please don't be intimidated by the
line count.

There are a number of interesting features:
- we infer the schema from type hints. In a followup I add the ability
  to manually specify a schema.
- we infer the namespace for the operator from the module. In a
  followup, we'll add the ability to manually specify a namespace.
- def_blackbox registrations can override each other. This makes them
  more pleasant to work with in environments like colab.
- we require that the outputs of the blackbox op do not alias any inputs
  or each other. We enforce this via a runtime check.

Test Plan:
- new tests

[ghstack-poisoned]

zou3519 committed:
Update on "Add torch.library.def_blackbox"


This is the entrypoint for defining a blackbox (e.g. PyTorch will
never peek into it) custom op. In this PR, you can specify backend impls
and the abstract impl for this op.

NB: most of this PR is docstrings, please don't be intimidated by the
line count.

There are a number of interesting features:
- we infer the schema from type hints. In a followup I add the ability
  to manually specify a schema.
- we infer the namespace for the operator from the module. In a
  followup, we'll add the ability to manually specify a namespace.
- def_blackbox registrations can override each other. This makes them
  more pleasant to work with in environments like colab.
- we require that the outputs of the blackbox op do not alias any inputs
  or each other. We enforce this via a runtime check.

Test Plan:
- new tests

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

zou3519 commented on test/test_custom_ops.py:
```diff
@@ -19,9 +19,11 @@
 from torch import Tensor
 from torch._custom_op.impl import custom_op, CustomOp, infer_schema
 from torch._utils_internal import get_file_path_2
+from torch.library import def_blackbox
```
Taking bikeshedding. `def_opaque`? Or just call it `@custom_op`? (We expect the majority of custom ops that people write will be blackbox).

zou3519 replied:
opaque/traceable sound more like opposites than blackbox/traceable

soulitzer replied:
> Or just call it @custom_op? (We expect the majority of custom ops that people write will be blackbox).

I like this idea, maybe custom_op and traceable_custom_op? (Maybe i'm biased because we've been referring to it as custom op all this time, but if someone does `@custom_op` it feels clearer to me that we're creating a custom op. )

zou3519 replied:
@soulitzer the general problem is that no all custom ops are opaque. If we want this to be clearer we could call the decorator opaque_op/traceable_op or blackbox_op/traceable_op instead of def_blackbox/def_traceable

zou3519 committed:
Update on "Add torch.library.def_blackbox"


Design doc: https://docs.google.com/document/d/13CS7nOEtTEsJ_0tiVGPhFDeYRbVb7LXIhw8atS-0eow/edit
Prototype PR: https://github.com/pytorch/pytorch/pull/120345

This is the entrypoint for defining a blackbox (e.g. PyTorch will
never peek into it) custom op. In this PR, you can specify backend impls
and the abstract impl for this op.

NB: most of this PR is docstrings, please don't be intimidated by the
line count.

There are a number of interesting features:
- we infer the schema from type hints. In a followup I add the ability
  to manually specify a schema.
- we infer the namespace for the operator from the module. In a
  followup, we'll add the ability to manually specify a namespace.
- def_blackbox registrations can override each other. This makes them
  more pleasant to work with in environments like colab.
- we require that the outputs of the blackbox op do not alias any inputs
  or each other. We enforce this via a runtime check.

Test Plan:
- new tests

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

zou3519 committed:
Update on "Add torch.library.def_blackbox"


Design doc: https://docs.google.com/document/d/13CS7nOEtTEsJ_0tiVGPhFDeYRbVb7LXIhw8atS-0eow/edit
Prototype PR: https://github.com/pytorch/pytorch/pull/120345

This is the entrypoint for defining a blackbox (e.g. PyTorch will
never peek into it) custom op. In this PR, you can specify backend impls
and the abstract impl for this op.

NB: most of this PR is docstrings, please don't be intimidated by the
line count.

There are a number of interesting features:
- we infer the schema from type hints. In a followup I add the ability
  to manually specify a schema.
- we infer the namespace for the operator from the module. In a
  followup, we'll add the ability to manually specify a namespace.
- def_blackbox registrations can override each other. This makes them
  more pleasant to work with in environments like colab.
- we require that the outputs of the blackbox op do not alias any inputs
  or each other. We enforce this via a runtime check.

Test Plan:
- new tests

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

zou3519 committed:
Update on "Add torch.library.def_blackbox"


Design doc: https://docs.google.com/document/d/13CS7nOEtTEsJ_0tiVGPhFDeYRbVb7LXIhw8atS-0eow/edit
Prototype PR: https://github.com/pytorch/pytorch/pull/120345

This is the entrypoint for defining a blackbox (e.g. PyTorch will
never peek into it) custom op. In this PR, you can specify backend impls
and the abstract impl for this op.

NB: most of this PR is docstrings, please don't be intimidated by the
line count.

There are a number of interesting features:
- we infer the schema from type hints. In a followup I add the ability
  to manually specify a schema.
- we infer the namespace for the operator from the module. In a
  followup, we'll add the ability to manually specify a namespace.
- def_blackbox registrations can override each other. This makes them
  more pleasant to work with in environments like colab.
- we require that the outputs of the blackbox op do not alias any inputs
  or each other. We enforce this via a runtime check.

Test Plan:
- new tests

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

ezyang commented on torch/_library/define.py:
```diff
@@ -0,0 +1,332 @@
+    """Wraps a function into a custom operator.
+
+    Treats the function as a black-box (that is, PyTorch will never peek into
+    the function). The function must have type hints; these are needed
```
This description doesn't seem clear. In particular, suppose I am a clueless user. Why would I not want PyTorch to peek into the function?

ezyang commented on test/test_custom_ops.py:
```diff
@@ -678,6 +680,134 @@ def foo(x: Tensor, y: int) -> Tensor:
+    def test_blackbox_basic(self):
+        @def_blackbox(mutated_args=())
+        def f(x: Tensor, y: float) -> Tensor:
+            return x + y
```
These examples are awkward because there isn't actually any reason for this operator to be black box

ezyang commented on test/test_custom_ops.py:
```diff
@@ -678,6 +680,134 @@ def foo(x: Tensor, y: int) -> Tensor:
+
+        @f.impl("cpu")
+        def _(x, y):
+            return x - y
```
And it's weird to have implementations that give numerically different results, that's not a thing you want people to do in reality.

ezyang commented on test/test_custom_ops.py:
```diff
@@ -678,6 +680,134 @@ def foo(x: Tensor, y: int) -> Tensor:
+        out = torch.compile(custom_linear, backend="eager", fullgraph=True)(
+            x, weight, bias
+        )
+        assert torch.allclose(out, torch.nn.functional.linear(x, weight, bias))
```
Use self.assertEqual? Also, if def_blackbox had done literally nothing, you would still pass this test. You probably need to toggle a flag inside the implementations to test that you've actually exercised the code in question.

zou3519 committed:
Update on "Add torch.library.opaque_op"


This is the entrypoint for defining an opaque/blackbox (e.g. PyTorch will
never peek into it) custom op. In this PR, you can specify backend impls
and the abstract impl for this op.

NB: most of this PR is docstrings, please don't be intimidated by the
line count.

There are a number of interesting features:
- we infer the schema from type hints. In a followup I add the ability
  to manually specify a schema.
- we infer the namespace for the operator from the module. In a
  followup, we'll add the ability to manually specify a namespace.
- opaque_op registrations can override each other. This makes them
  more pleasant to work with in environments like colab.
- we require that the outputs of the opaque op do not alias any inputs
  or each other. We enforce this via a runtime check.

Test Plan:
- new tests

[ghstack-poisoned]

ezyang commented on test/test_custom_ops.py:
```diff
@@ -678,6 +680,134 @@ def foo(x: Tensor, y: int) -> Tensor:
+        y = f(x)
+        self.assertEqual(y, x.sin())
+
+    def test_blackbox_disallows_output_aliasing(self):
```
Huh, so is the test done at runtime? Seems like it could potentially be a source of overhead

zou3519 replied:
Yes, the test is done at runtime. We could add it to opcheck instead to avoid overhead.

zou3519 replied:
Btw, autograd.Function does similar checks at runtime, so, I'm not worried about performance right now. If it becomes a problem we can move the check to opcheck instead of runtime.

ezyang commented on torch/_library/define.py:
```diff
@@ -0,0 +1,334 @@
+    assert len(mutated_args) == 0, "NYI"
+    mod = get_caller_module(stacklevel=2)
+    if mod is None:
+        namespace = "module_not_found"  # can happen in colab
```
Probably need to detect conflicts in this case

ezyang commented on torch/_library/define.py:
```diff
@@ -0,0 +1,334 @@
+
+
+def mangle_module(module):
+    """Mangles the module name.
```
So... do you have an unmangler? Because that's the point of mangling

fwiw, I would suggest reserving __ as the dot separator, and then doing an escaping scheme for preexisting `__` in the name. Done this way, you don't need to have a `mangledN` signifier, the mangling is always the same.

ezyang commented on torch/_library/define.py:
```diff
@@ -0,0 +1,334 @@
+    """Returns the fully qualified name of the module that called this function."""
+    frame = sys._getframe(stacklevel)
+    mod = inspect.getmodule(frame)
+    return mod
```
How come we can't just read the module off of the function that the decorator is called with?

zou3519 replied:
We can! thanks for pointing it out.

ezyang commented on torch/_library/define.py:
```diff
@@ -0,0 +1,334 @@
+
+
+class OpaqueOpDef:
+    def __init__(self, namespace: str, name: str, schema: str):
```
Consider adding return type annots here and every where else

zou3519 replied:
done

ezyang commented on torch/_library/define.py:
```diff
@@ -0,0 +1,334 @@
+
+                raise torch._subclasses.fake_tensor.UnsupportedOperatorException(
+                    self._opoverload
+                )
```
It seems like you should explicitly suggest the user add an `impl_abstract` to their operator definition when this is hit. In fact, I would probably advise against Dynamo automatically graph breaking in this situation (because that's what the user would have gotten when they didn't black box the op).

ezyang commented on torch/_library/define.py:
```diff
@@ -0,0 +1,334 @@
+
+                    def backend_impl(*args, **kwargs):
+                        # Checks the assumption that outputs cannot alias
+                        # inputs or other outputs.
```
yup runtime overhead

ezyang commented on torch/_library/define.py:
```diff
@@ -0,0 +1,334 @@
+                    def backend_impl(*args, **kwargs):
+                        # Checks the assumption that outputs cannot alias
+                        # inputs or other outputs.
+                        storages = set({})
```
nit: set() is fine

ezyang commented on torch/_library/define.py:
```diff
@@ -0,0 +1,334 @@
+    qualname = f"{namespace}::{name}"
+
+    if qualname in OPDEF_TO_LIB:
+        OPDEF_TO_LIB[qualname]._destroy()
```
🤔

ezyang replied:
Is the reason you do this because, even though you mangle modules, you absolutely do not want to mangle names? But if the modules are mangled, there isn't really any expectation of name stability, and I feel you should be willing to mangle names.

If someone wants to rely on the exact naming of the custom operator, then we should force the user to explicitly say what they want the name to be. In fact, it may be worthwhile to make it a lot *harder* for people to try to manually reverse engineer the mangling scheme, because the current module mangling means that transforms like moving a function from one file to another will change its name and would then be BC-breaking. One way to make it hard is to put a big ass string in the module name being like "DONT_USE_THIS_GIVE_EXPLICIT_NAME_IF_NEEDED"

zou3519 replied:
Updated the mangling scheme as per your comments

zou3519 committed:
Update on "Add torch.library.opaque_op"


This is the entrypoint for defining an opaque/blackbox (e.g. PyTorch will
never peek into it) custom op. In this PR, you can specify backend impls
and the abstract impl for this op.

NB: most of this PR is docstrings, please don't be intimidated by the
line count.

There are a number of interesting features:
- we infer the schema from type hints. In a followup I add the ability
  to manually specify a schema.
- we infer the namespace for the operator from the module. In a
  followup, we'll add the ability to manually specify a namespace.
- opaque_op registrations can override each other. This makes them
  more pleasant to work with in environments like colab.
- we require that the outputs of the opaque op do not alias any inputs
  or each other. We enforce this via a runtime check.

Test Plan:
- new tests

[ghstack-poisoned]

zou3519 committed:
Update on "Add torch.library.custom_op"


This is the entrypoint for defining an opaque/blackbox (e.g. PyTorch will
never peek into it) custom op. In this PR, you can specify backend impls
and the abstract impl for this op.

NB: most of this PR is docstrings, please don't be intimidated by the
line count.

There are a number of interesting features:
- we infer the schema from type hints. In a followup I add the ability
  to manually specify a schema.
- we infer the qualname for the operator from the provided function. In a
  followup, we'll add the ability to manually specify a qualname.
- custom_op registrations can override each other. This makes them
  more pleasant to work with in environments like colab.
- we require that the outputs of the custom_op do not alias any inputs
  or each other. We enforce this via a runtime check, but can relax this
  into an opcheck test if it really matters in the future.

Test Plan:
- new tests

[ghstack-poisoned]

zou3519 committed:
Update on "Add torch.library.custom_op"


This is the entrypoint for defining an opaque/blackbox (e.g. PyTorch will
never peek into it) custom op. In this PR, you can specify backend impls
and the abstract impl for this op.

NB: most of this PR is docstrings, please don't be intimidated by the
line count.

There are a number of interesting features:
- we infer the schema from type hints. In a followup I add the ability
  to manually specify a schema.
- we infer the qualname for the operator from the provided function. In a
  followup, we'll add the ability to manually specify a qualname.
- custom_op registrations can override each other. This makes them
  more pleasant to work with in environments like colab.
- we require that the outputs of the custom_op do not alias any inputs
  or each other. We enforce this via a runtime check, but can relax this
  into an opcheck test if it really matters in the future.

Test Plan:
- new tests

[ghstack-poisoned]

zou3519 commented:
This should be ready for another review. TL;DR of changes:
- naming changes per user/dev feedback
  - def_blackbox -> custom_op
  - impl -> register_impl
  - impl_abstract -> register_fake
- changed the mangling scheme as per comments

There are a number of TODOs in the code that I'd prefer to do as followups (this PR is getting long), but please lmk if you'd prefer any to be handled now.

albanD commented on torch/library.py:
```diff
@@ -178,6 +185,8 @@ def _destroy(self):
             handle.destroy()
         self._registration_handles.clear()
+        global _impls
+        _impls -= self._op_impls
```
That is a current bug?

zou3519 replied:
Yes

albanD commented on torch/_library/custom_ops.py:
```diff
@@ -0,0 +1,363 @@
+            >>>
+            >>> # Add implementations for the cuda device
+            >>> @numpy_sin.register_impl("cuda")
+            >>> def _(x):
```
tbh I think this will lead to more questions than anything else.
I think we should use a name that makes sense here or numpy_sin like is done with python property.

zou3519 replied:
>  or numpy_sin like is done with python property.

What do you mean?

albanD replied:
The default example for property from python is to name all of these the same: https://docs.python.org/3/library/functions.html#property.deleter

zou3519 replied:
I need to go think about this a bit more

zou3519 replied:
We've got four options:
```py
@custom_op("mylib::sin", mutated_args=())
def sin(x):
    return x.clone()

# Option 1: recommend underscore name (The python singledispatch way)
# https://docs.python.org/3/library/functools.html#functools.singledispatch
@sin.register_fake
def _(x):
    return torch.empty_like(x)

# Option 2: recommend a named name
@sin.register_fake
def sin_fake(x):
    return torch.empty_like(x)

# Option 3: recommend the original name (The python property way)
@sin.register_fake
def sin(x):
    return torch.empty_like(x)

# Option 4: recommend a random name with convention
@sin.register_fake
def impl(x):
    return torch.empty_like(x)
```

I don't like Option 2 and Option 3 because those have bitten me in the past:
- Option 2 leads to unnecessarily verbose names and copy-paste confusion
- Option 3 means we can't call any of the impls directly.

ezyang replied:
I don't really have a bone in this fight but Python singledispatch following the _ convention seems pretty compelling to me

zou3519 commented on torch/_library/custom_ops.py:
```diff
@@ -0,0 +1,363 @@
+    Args:
+        mutated_args (Sequence[str]): The names of args that the function mutates.
+            This MUST be accurate, otherwise, the behavior is undefined.
+        types (None | str | Sequence[str]): The device type(s) the function
```
It's not exactly a device (devices have indices). We can call it `device_types`?

albanD replied:
device_types sounds good if we want to be exact.

zou3519 commented on torch/_library/custom_ops.py:
```diff
@@ -0,0 +1,363 @@
+        self._build()
+
+    @classmethod
+    def _from_fn(
```
I don't want users to rely on this

albanD replied:
But they can rely on the rest of the class here?

zou3519 replied:
Yup

zou3519 replied:
I moved this out to avoid confusion. but in general there are a lot of private attributes on this class (all prefixed with _) that I don't want people to rely on.

zou3519 commented on torch/_library/custom_ops.py:
```diff
@@ -0,0 +1,363 @@
+    ) -> Callable:
+        """Register an implementation for a device type for this operator.
+
+        Some valid types are: "cpu", "cuda", "xla", "mps", "ipu", "xpu".
```
Nope :P

zou3519 commented on torch/_library/custom_ops.py:
```diff
@@ -0,0 +1,363 @@
+
+        if fn is None:
+            return inner
+        return inner(fn)
```
The register_ decorators should return the original function. This allows for things like double decoration and calling the original function directly.

zou3519 commented on torch/_library/custom_ops.py:
```diff
@@ -0,0 +1,363 @@
+OPDEF_TO_LIB: Dict[str, "library.Library"] = {}
+
+
+def get_library_allowing_overwrite(namespace: str, name: str) -> "library.Library":
```
This will delete the old library, which will delete all registrations associated with that library.

albanD replied:
Not sure how this will play with C++ implemented kernels. But sounds ok.

zou3519 replied:
We don't need to worry about that for now.

zou3519 commented on torch/_library/custom_ops.py:
```diff
@@ -0,0 +1,363 @@
+                self._backend_fns[device_type] = fn
+            return fn
+
+        if fn is None:
```
if fn is None, then this API is being used as a decorator

albanD replied:
Ho right, a comment explaining this would be nice :)

albanD commented on torch/_library/custom_ops.py:
```diff
@@ -0,0 +1,363 @@
+
+        """
+        self._abstract_fn = fn
+        return fn
```
Ho I though that's what the call in init `self._build()` is?

zou3519 replied:
That only gets called once, when the user calls `@custom_op`

zou3519 committed:
Update on "Add torch.library.custom_op"


This is the entrypoint for defining an opaque/blackbox (e.g. PyTorch will
never peek into it) custom op. In this PR, you can specify backend impls
and the abstract impl for this op.

NB: most of this PR is docstrings, please don't be intimidated by the
line count.

There are a number of interesting features:
- we infer the schema from type hints. In a followup I add the ability
  to manually specify a schema.
- name inference. The user needs to manually specify an op name for now.
  In a followup we add the ability to automatically infer a name (this
  is a little tricky).
- custom_op registrations can override each other. This makes them
  more pleasant to work with in environments like colab.
- we require that the outputs of the custom_op do not alias any inputs
  or each other. We enforce this via a runtime check, but can relax this
  into an opcheck test if it really matters in the future.

Test Plan:
- new tests

[ghstack-poisoned]

zou3519 committed:
Update on "Add torch.library.custom_op"


This is the entrypoint for defining an opaque/blackbox (e.g. PyTorch will
never peek into it) custom op. In this PR, you can specify backend impls
and the abstract impl for this op.

NB: most of this PR is docstrings, please don't be intimidated by the
line count.

There are a number of interesting features:
- we infer the schema from type hints. In a followup I add the ability
  to manually specify a schema.
- name inference. The user needs to manually specify an op name for now.
  In a followup we add the ability to automatically infer a name (this
  is a little tricky).
- custom_op registrations can override each other. This makes them
  more pleasant to work with in environments like colab.
- we require that the outputs of the custom_op do not alias any inputs
  or each other. We enforce this via a runtime check, but can relax this
  into an opcheck test if it really matters in the future.

Test Plan:
- new tests

[ghstack-poisoned]

zou3519 committed:
Update on "Add torch.library.custom_op"


This is the entrypoint for defining an opaque/blackbox (e.g. PyTorch will
never peek into it) custom op. In this PR, you can specify backend impls
and the abstract impl for this op.

NB: most of this PR is docstrings, please don't be intimidated by the
line count.

There are a number of interesting features:
- we infer the schema from type hints. In a followup I add the ability
  to manually specify a schema.
- name inference. The user needs to manually specify an op name for now.
  In a followup we add the ability to automatically infer a name (this
  is a little tricky).
- custom_op registrations can override each other. This makes them
  more pleasant to work with in environments like colab.
- we require that the outputs of the custom_op do not alias any inputs
  or each other. We enforce this via a runtime check, but can relax this
  into an opcheck test if it really matters in the future.

Test Plan:
- new tests

[ghstack-poisoned]

zou3519 commented:
High-level updates:
1. I have split off the automatic namespace inference to a future PR, to not complicate this one (and because there's a lot of parallel work I want to do that is blocked on this PR). Instead, we require the user to pass in a manual name for their custom op in this PR
2. Updated with Alban's feedback

ezyang commented on torch/_library/custom_ops.py:
```diff
@@ -0,0 +1,349 @@
+
+def custom_op(
+    name: str,
+    /,
```
aw yisss

ezyang commented on torch/_library/custom_ops.py:
```diff
@@ -0,0 +1,349 @@
+    return lib
+
+
+def iter_tensors(
```
Intended to be private?

zou3519 committed:
Update on "Add torch.library.custom_op"


This is the entrypoint for defining an opaque/blackbox (e.g. PyTorch will
never peek into it) custom op. In this PR, you can specify backend impls
and the abstract impl for this op.

NB: most of this PR is docstrings, please don't be intimidated by the
line count.

There are a number of interesting features:
- we infer the schema from type hints. In a followup I add the ability
  to manually specify a schema.
- name inference. The user needs to manually specify an op name for now.
  In a followup we add the ability to automatically infer a name (this
  is a little tricky).
- custom_op registrations can override each other. This makes them
  more pleasant to work with in environments like colab.
- we require that the outputs of the custom_op do not alias any inputs
  or each other. We enforce this via a runtime check, but can relax this
  into an opcheck test if it really matters in the future.

Test Plan:
- new tests

[ghstack-poisoned]

zou3519 committed:
Update on "Add torch.library.custom_op"


This is the entrypoint for defining an opaque/blackbox (e.g. PyTorch will
never peek into it) custom op. In this PR, you can specify backend impls
and the abstract impl for this op.

NB: most of this PR is docstrings, please don't be intimidated by the
line count.

There are a number of interesting features:
- we infer the schema from type hints. In a followup I add the ability
  to manually specify a schema.
- name inference. The user needs to manually specify an op name for now.
  In a followup we add the ability to automatically infer a name (this
  is a little tricky).
- custom_op registrations can override each other. This makes them
  more pleasant to work with in environments like colab.
- we require that the outputs of the custom_op do not alias any inputs
  or each other. We enforce this via a runtime check, but can relax this
  into an opcheck test if it really matters in the future.

Test Plan:
- new tests

[ghstack-poisoned]

albanD commented on torch/library.py:
```diff
@@ -16,6 +18,7 @@
     'fallthrough_kernel',
     'impl_abstract',
     'get_ctx',
+    'custom_op',
```
I'm surprised this is not tripping the public API checks. The `__module__` being private shouldn't be ok...

albanD replied:
It should definitely fail: https://github.com/pytorch/pytorch/blob/72662bf05b3499ce96aae9183a489c78f0c44c84/test/test_public_bindings.py#L477
Also because of this, we miss the fact that this is not properly documented on the website.

zou3519 replied:
Lol we need to update the public bindings test to catch these cases

albanD commented on torch/_library/custom_ops.py:
```diff
@@ -0,0 +1,351 @@
+    def check(arg):
+        if isinstance(arg, Tensor):
+            yield arg
+        elif allowed_nesting > 0 and isinstance(arg, (tuple, list)):
```
nit: typing says Tuple and this says tuple/list?
Should this just be Sequence?

zou3519 replied:
The type hint says that `args` is a `Tuple[Any]`. `arg` is `Any` (and it may be a tuple or a list)

albanD commented on torch/_library/custom_ops.py:
```diff
@@ -0,0 +1,351 @@
+
+
+def iter_tensors(
+    args: Tuple[Any], kwargs: Dict[str, Any], allowed_nesting: int = 1
```
Should we remove that allowed_nesting arg since it is never used?

zou3519 replied:
That's a bug, thanks for the catch

zou3519 replied:
Actually, I see what's going on. This function is recursive. Inside `iter_tensors`, I call `iter_tensors` with a different allowed_nesting.

albanD commented on torch/_library/custom_ops.py:
```diff
@@ -0,0 +1,351 @@
+        self._abstract_fn = fn
+        return fn
+
+    def _register_to_dispatcher(self) -> None:
```
nit: "_define_in_dispatcher" ?

zou3519 replied:
Well, we technically define + impl in this function

albanD reviewed (approved):
Sounds good! Just a few nits/questions that can be done in follow ups

albanD commented on torch/_library/custom_ops.py:
```diff
@@ -0,0 +1,351 @@
+        # Fields used to interface with the PyTorch dispatcher
+        self._namespace = namespace
+        self._name = name
+        self._qualname = f"{self._namespace}::{self._name}"
```
nit: make this a property computed dynamically removes the two source of truth and avoid potential divergence

zou3519 replied:
yessir

zou3519 committed:
Update on "Add torch.library.custom_op"


This is the entrypoint for defining an opaque/blackbox (e.g. PyTorch will
never peek into it) custom op. In this PR, you can specify backend impls
and the abstract impl for this op.

NB: most of this PR is docstrings, please don't be intimidated by the
line count.

There are a number of interesting features:
- we infer the schema from type hints. In a followup I add the ability
  to manually specify a schema.
- name inference. The user needs to manually specify an op name for now.
  In a followup we add the ability to automatically infer a name (this
  is a little tricky).
- custom_op registrations can override each other. This makes them
  more pleasant to work with in environments like colab.
- we require that the outputs of the custom_op do not alias any inputs
  or each other. We enforce this via a runtime check, but can relax this
  into an opcheck test if it really matters in the future.

Test Plan:
- new tests

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

zou3519 committed:
Update on "Add torch.library.custom_op"


This is the entrypoint for defining an opaque/blackbox (e.g. PyTorch will
never peek into it) custom op. In this PR, you can specify backend impls
and the abstract impl for this op.

NB: most of this PR is docstrings, please don't be intimidated by the
line count.

There are a number of interesting features:
- we infer the schema from type hints. In a followup I add the ability
  to manually specify a schema.
- name inference. The user needs to manually specify an op name for now.
  In a followup we add the ability to automatically infer a name (this
  is a little tricky).
- custom_op registrations can override each other. This makes them
  more pleasant to work with in environments like colab.
- we require that the outputs of the custom_op do not alias any inputs
  or each other. We enforce this via a runtime check, but can relax this
  into an opcheck test if it really matters in the future.

Test Plan:
- new tests

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]
